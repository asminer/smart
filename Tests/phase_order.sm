


/*
  Checks order statistics for phase types
*/

/*
  Example from lecture notes...
*/

dtmc mygeom := {
  state s1, s0;
  init (s1:1);
  arcs(
    s1:s1:3, s1:s0:1
  );

  ph int make := tta(in_state(s0));
};

dtmc myequi := {
  state s3, s2, s1, s0;
  init(s3:1, s2:1, s1:1);
  arcs(
    s3:s2:1, s2:s1:1, s1:s0:1
  );
  ph int make := tta(in_state(s0));
};

ph int g14 := mygeom.make;
ph int eq13 := myequi.make;

print_ph(g14);
print_ph(eq13);

// example 27.10

print_ph(max(g14, eq13));

// example 21.11

print_ph(min(g14, eq13));


/*
  Similar examples, but using our phase representation
*/

print_ph(geometric(0.25));
print_ph(equilikely(1, 3));

print_ph(max(geometric(0.25), equilikely(1, 3)));
print_ph(min(geometric(0.25), equilikely(1, 3)));


/*
  Some easy, multi-dimensional, test cases
*/

ph int one := 1;
ph int two := 2;
ph int three := 3;
ph int four := 4;

print_ph(max(one, two, three, four));
print_ph(min(one, two, three, four));

/*
  Continuous tests
*/

ph real Ca := expo(3.5);
ph real Cb := expo(2.0);

print_ph(max(Ca, Cb));
print_ph(min(Ca, Cb));
// note we get memoryless for free here!

print_ph(max(erlang(2, 2.2), erlang(3, 3.3)));
print_ph(min(erlang(2, 2.2), erlang(3, 3.3)));

