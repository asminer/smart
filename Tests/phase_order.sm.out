Information for phase type tta(mygeom, ...)
Discrete phase type, state has dimension 1
Reachable states:
State 0: 0
State 1: a
2 states total
Accepting state index: 1
No trap state
Initial distribution:
    (0:1)
Markov chain:
From state 0:
	To state 0 with weight 0.75
	To state 1 with weight 0.25
From state 1:
	To state 1 with weight 1
3 edges total
End of information for phase type

Information for phase type tta(myequi, ...)
Discrete phase type, state has dimension 1
Reachable states:
State 0: 0
State 1: 1
State 2: 2
State 3: a
4 states total
Accepting state index: 3
No trap state
Initial distribution:
    (0:0.333333, 1:0.333333, 2:0.333333)
Markov chain:
From state 0:
	To state 1 with weight 1
From state 1:
	To state 2 with weight 1
From state 2:
	To state 3 with weight 1
From state 3:
	To state 3 with weight 1
4 edges total
End of information for phase type

Information for phase type max(tta(mygeom, ...), tta(myequi, ...))
Discrete phase type, state has dimension 2
Reachable states:
State 0: [0, 0]
State 1: [0, 1]
State 2: [0, 2]
State 3: [a, 1]
State 4: [a, 2]
State 5: [0, a]
State 6: [a, a]
7 states total
Accepting state index: 6
No trap state
Initial distribution:
    (0:0.333333, 1:0.333333, 2:0.333333)
Markov chain:
From state 0:
	To state 1 with weight 0.75
	To state 3 with weight 0.25
From state 1:
	To state 2 with weight 0.75
	To state 4 with weight 0.25
From state 2:
	To state 5 with weight 0.75
	To state 6 with weight 0.25
From state 3:
	To state 4 with weight 1
From state 4:
	To state 6 with weight 1
From state 5:
	To state 5 with weight 0.75
	To state 6 with weight 0.25
From state 6:
	To state 6 with weight 1
11 edges total
End of information for phase type

Information for phase type min(tta(mygeom, ...), tta(myequi, ...))
Discrete phase type, state has dimension 2
Reachable states:
State 0: [0, 0]
State 1: [0, 1]
State 2: [0, 2]
State 3: [a, a]
4 states total
Accepting state index: 3
No trap state
Initial distribution:
    (0:0.333333, 1:0.333333, 2:0.333333)
Markov chain:
From state 0:
	To state 1 with weight 0.75
	To state 3 with weight 0.25
From state 1:
	To state 2 with weight 0.75
	To state 3 with weight 0.25
From state 2:
	To state 3 with weight 1
From state 3:
	To state 3 with weight 1
6 edges total
End of information for phase type

Information for phase type Geometric(0.25)
Discrete phase type, state has dimension 1
Reachable states:
State 0: 1
State 1: a
2 states total
Accepting state index: 1
No trap state
Initial distribution:
    (0:0.25, 1:0.75)
Markov chain:
From state 0:
	To state 0 with weight 0.25
	To state 1 with weight 0.75
From state 1:
	To state 1 with weight 1
3 edges total
End of information for phase type

Information for phase type Equilikely(1, 3)
Discrete phase type, state has dimension 1
Reachable states:
State 0: 1
State 1: 2
State 2: 3
State 3: a
4 states total
Accepting state index: 3
No trap state
Initial distribution:
    (0:1)
Markov chain:
From state 0:
	To state 1 with weight 0.666667
	To state 3 with weight 0.333333
From state 1:
	To state 2 with weight 0.5
	To state 3 with weight 0.5
From state 2:
	To state 3 with weight 1
From state 3:
	To state 3 with weight 1
6 edges total
End of information for phase type

Information for phase type max(Geometric(0.25), Equilikely(1, 3))
Discrete phase type, state has dimension 2
Reachable states:
State 0: [1, 1]
State 1: [a, 1]
State 2: [1, 2]
State 3: [a, 2]
State 4: [1, a]
State 5: [1, 3]
State 6: [a, 3]
State 7: [a, a]
8 states total
Accepting state index: 7
No trap state
Initial distribution:
    (0:0.25, 1:0.75)
Markov chain:
From state 0:
	To state 2 with weight 0.166667
	To state 3 with weight 0.5
	To state 4 with weight 0.0833333
	To state 7 with weight 0.25
From state 1:
	To state 3 with weight 0.666667
	To state 7 with weight 0.333333
From state 2:
	To state 4 with weight 0.125
	To state 5 with weight 0.125
	To state 6 with weight 0.375
	To state 7 with weight 0.375
From state 3:
	To state 6 with weight 0.5
	To state 7 with weight 0.5
From state 4:
	To state 4 with weight 0.25
	To state 7 with weight 0.75
From state 5:
	To state 4 with weight 0.25
	To state 7 with weight 0.75
From state 6:
	To state 7 with weight 1
From state 7:
	To state 7 with weight 1
18 edges total
End of information for phase type

Information for phase type min(Geometric(0.25), Equilikely(1, 3))
Discrete phase type, state has dimension 2
Reachable states:
State 0: [1, 1]
State 1: [1, 2]
State 2: [1, 3]
State 3: [a, a]
4 states total
Accepting state index: 3
No trap state
Initial distribution:
    (0:0.25, 3:0.75)
Markov chain:
From state 0:
	To state 1 with weight 0.166667
	To state 3 with weight 0.833333
From state 1:
	To state 2 with weight 0.125
	To state 3 with weight 0.875
From state 2:
	To state 3 with weight 1
From state 3:
	To state 3 with weight 1
6 edges total
End of information for phase type

Information for phase type max(1, 2, 3, 4)
Discrete phase type, state has dimension 4
Reachable states:
State 0: [1, 1, 1, 1]
State 1: [a, 2, 2, 2]
State 2: [a, a, 3, 3]
State 3: [a, a, a, 4]
State 4: [a, a, a, a]
5 states total
Accepting state index: 4
No trap state
Initial distribution:
    (0:1)
Markov chain:
From state 0:
	To state 1 with weight 1
From state 1:
	To state 2 with weight 1
From state 2:
	To state 3 with weight 1
From state 3:
	To state 4 with weight 1
From state 4:
	To state 4 with weight 1
5 edges total
End of information for phase type

Information for phase type min(1, 2, 3, 4)
Discrete phase type, state has dimension 4
Reachable states:
State 0: [1, 1, 1, 1]
State 1: [a, a, a, a]
2 states total
Accepting state index: 1
No trap state
Initial distribution:
    (0:1)
Markov chain:
From state 0:
	To state 1 with weight 1
From state 1:
	To state 1 with weight 1
2 edges total
End of information for phase type

Information for phase type max(expo(3.5), expo(2))
Continuous phase type, state has dimension 2
Reachable states:
State 0: [0, 0]
State 1: [a, 0]
State 2: [0, a]
State 3: [a, a]
4 states total
Accepting state index: 3
No trap state
Initial distribution:
    (0:1)
Markov chain:
From state 0:
	To state 1 with weight 3.5
	To state 2 with weight 2
From state 1:
	To state 3 with weight 2
From state 2:
	To state 3 with weight 3.5
From state 3:
4 edges total
End of information for phase type

Information for phase type min(expo(3.5), expo(2))
Continuous phase type, state has dimension 2
Reachable states:
State 0: [0, 0]
State 1: [a, a]
2 states total
Accepting state index: 1
No trap state
Initial distribution:
    (0:1)
Markov chain:
From state 0:
	To state 1 with weight 5.5
From state 1:
1 edges total
End of information for phase type

Information for phase type max(erlang(2, 2.2), erlang(3, 3.3))
Continuous phase type, state has dimension 2
Reachable states:
State 0: [0, 0]
State 1: [1, 0]
State 2: [0, 1]
State 3: [a, 0]
State 4: [1, 1]
State 5: [0, 2]
State 6: [a, 1]
State 7: [1, 2]
State 8: [0, a]
State 9: [a, 2]
State 10: [1, a]
State 11: [a, a]
12 states total
Accepting state index: 11
No trap state
Initial distribution:
    (0:1)
Markov chain:
From state 0:
	To state 1 with weight 2.2
	To state 2 with weight 3.3
From state 1:
	To state 3 with weight 2.2
	To state 4 with weight 3.3
From state 2:
	To state 4 with weight 2.2
	To state 5 with weight 3.3
From state 3:
	To state 6 with weight 3.3
From state 4:
	To state 6 with weight 2.2
	To state 7 with weight 3.3
From state 5:
	To state 7 with weight 2.2
	To state 8 with weight 3.3
From state 6:
	To state 9 with weight 3.3
From state 7:
	To state 9 with weight 2.2
	To state 10 with weight 3.3
From state 8:
	To state 10 with weight 2.2
From state 9:
	To state 11 with weight 3.3
From state 10:
	To state 11 with weight 2.2
From state 11:
17 edges total
End of information for phase type

Information for phase type min(erlang(2, 2.2), erlang(3, 3.3))
Continuous phase type, state has dimension 2
Reachable states:
State 0: [0, 0]
State 1: [1, 0]
State 2: [0, 1]
State 3: [1, 1]
State 4: [0, 2]
State 5: [1, 2]
State 6: [a, a]
7 states total
Accepting state index: 6
No trap state
Initial distribution:
    (0:1)
Markov chain:
From state 0:
	To state 1 with weight 2.2
	To state 2 with weight 3.3
From state 1:
	To state 3 with weight 3.3
	To state 6 with weight 2.2
From state 2:
	To state 3 with weight 2.2
	To state 4 with weight 3.3
From state 3:
	To state 5 with weight 3.3
	To state 6 with weight 2.2
From state 4:
	To state 5 with weight 2.2
	To state 6 with weight 3.3
From state 5:
	To state 6 with weight 5.5
From state 6:
11 edges total
End of information for phase type

